build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/vagrant/VagrantShellTask.java

```diff
@@ -0,0 +1,113 @@
+
+    private final VagrantExtension extension;
+    private final VagrantMachine service;
+    private UnaryOperator<String> progressHandler = UnaryOperator.identity();
+
+    public VagrantShellTask() {
+        extension = getProject().getExtensions().findByType(VagrantExtension.class);
+        if (extension == null) {
+            throw new IllegalStateException("elasticsearch.vagrant-base must be applied to create " + getClass().getName());
+        }
+        service = getProject().getExtensions().getByType(VagrantMachine.class);
+    }
+
+    @Input
+    protected abstract List<String> getWindowsScript();
+
+    @Input
+    protected abstract List<String> getLinuxScript();
+
+    @Input
+    public UnaryOperator<String> getProgressHandler() {
+        return progressHandler;
+    }
+
+    public void setProgressHandler(UnaryOperator<String> progressHandler) {
+        this.progressHandler = progressHandler;
+    }
+
+    @TaskAction
+    public void runScript() {
+        String rootDir = getProject().getRootDir().toString();
+        if (extension.isWindowsVM()) {
+            service.execute(spec -> {
+                spec.setCommand("winrm");
+
+                List<String> script = new ArrayList<>();
+                script.add("try {");
+                script.add("cd " + convertWindowsPath(getProject(), rootDir));
+                extension.getVmEnv().forEach((k, v) -> script.add("$Env:" + k + " = \"" + v + "\""));
+                script.addAll(getWindowsScript().stream().map(s -> "    " + s).collect(Collectors.toList()));
+                script.addAll(
+                    Arrays.asList(
+                        "    exit $LASTEXITCODE",
+                        "} catch {",
+                        // catch if we have a failure to even run the script at all above, equivalent to set -e, sort of
+                        "    echo $_.Exception.Message",
+                        "    exit 1",
+                        "}"
+                    )
+                );
+                spec.setArgs("--elevated", "--command", String.join("\n", script));
+                spec.setProgressHandler(progressHandler);
+            });
+        } else {
+            try {
+                service.execute(spec -> {
+                    spec.setCommand("ssh");
+
+                    List<String> script = new ArrayList<>();
+                    script.add("sudo bash -c '"); // start inline bash script
+                    script.add("pwd");
+                    script.add("cd " + convertLinuxPath(getProject(), rootDir));
+                    extension.getVmEnv().forEach((k, v) -> script.add("export " + k + "=" + v));
+                    script.addAll(getLinuxScript());
+                    script.add("'"); // end inline bash script
+                    spec.setArgs("--command", String.join("\n", script));
+                    spec.setProgressHandler(progressHandler);
+                });
+            } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)